src/network/clientpackethandler.cpp

```diff
@@ -1395,3 +1396,66 @@ void Client::handleCommand_CSMFlavourLimits(NetworkPacket *pkt)
 {
 	*pkt >> m_csm_flavour_limits >> m_csm_noderange_limit;
 }
+
+/*
+ * Mod channels
+ */
+
+// @TODO client should have a list of registered channels, to filter them directly in API
+// it should also verify if it joined channels here, and do some rate limit too
+void Client::handleCommand_ModChannelMsg(NetworkPacket *pkt)
+{
+	std::string channel_name, channel_msg;
+	*pkt >> channel_name >> channel_msg;
+
+	verbosestream << "Mod channel message received from server " << pkt->getPeerId()
+		<< " on channel " << channel_name << " message: " << channel_msg << std::endl;
+
+	if (!m_modchannel_mgr->channelRegistered(channel_name)) {
+		verbosestream << "Server sent us messages on unregistered channel "
+			<< channel_name << ", ignoring." << std::endl;
+		return;
+	}
+
+	m_script->on_modchannel_message(channel_name, channel_msg);
+}
+
+void Client::handleCommand_ModChannelSignal(NetworkPacket *pkt)
+{
+	u8 signal_tmp;
+	ModChannelSignal signal;
+	std::string channel;
+
+	*pkt >> signal_tmp >> channel;
+
+	signal = (ModChannelSignal)signal_tmp;
+	// @TODO: send Signal to Lua API
+	switch (signal) {
+		case MODCHANNEL_SIGNAL_JOIN_OK:
+			m_modchannel_mgr->joinChannel(channel, 0);
+			infostream << "Server ack our mod channel join on channel `" << channel
+				<< "`, joining." << std::endl;
+			break;
+		case MODCHANNEL_SIGNAL_JOIN_FAILURE:
+			infostream << "Server refused our mod channel join on channel `" << channel
+				<< "`" << std::endl;
+			break;
+		case MODCHANNEL_SIGNAL_LEAVE_OK:
+			m_modchannel_mgr->leaveChannel(channel, 0);
+			infostream << "Server ack our mod channel leave on channel " << channel
+				<< "`, leaving." << std::endl;
+			break;
+		case MODCHANNEL_SIGNAL_LEAVE_FAILURE:
```

Comment:
why would <i>leaving</i> a channel ever be refused?
why does this need acknowledgement at all?

Reply:
an implementation error for example, can cause this, example: a CSM mod asking for leaving on a non joined channel. I tend to prefer ack isntead of trust, as it's an asynchronous model

Reply:
i get that, but why does the client need to know?

sever says "you can't leave that because you didn't join", client is like "oh, okay" and does nothing at all